combat/CombatSystem.cpp

```diff
@@ -1303,7 +1301,7 @@ namespace {
         }
     };
 
-    std::vector<PartAttackInfo> GetWeapons(std::shared_ptr<UniverseObject>& attacker) {
+    std::vector<PartAttackInfo> GetWeapons(std::shared_ptr<UniverseObject> attacker) {
```

Comment:
Reason for passing by value and not reference?

Reply:
GetWeapons ist not called with shared_ptr<UniverseObject> all the time, but shared_ptr<Ship>.

I'm also not sure it we are even using the shared_ptr as intended.

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#S-resource
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-smartptrparam

Reply:
> GetWeapons ist not called with shared_ptr all the time, but shared_ptr.

?

Reply:
Fixed the comment. Github ate the angle brackets.

Reply:
> GetWeapons ist not called with shared_ptr<UniverseObject> all the time, but shared_ptr<Ship>.

Not sure of the relevance... pointer-to-derived should be passable as pointer-to-base either way.

My thinking is that passing by value, if not optimized away, would require a copy of the pointer, which involves incrementing the reference count, while passing by const reference would not.

Reply:
You don't pass a pointer-to-derived, but a shared_ptr<T> as shared_ptr<U>&, which are unrelated types.  The fact that T is-a U doesn't matter for the compiler here.